src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs

```diff
@@ -35,6 +40,581 @@ public byte[] Deserialize(ref MessagePackReader reader, MessagePackSerializerOpt
         }
     }
 
+    public sealed class SByteArrayFormatter : IMessagePackFormatter<sbyte[]>
+    {
+        public static readonly SByteArrayFormatter Instance = new SByteArrayFormatter();
+
+        private SByteArrayFormatter()
+        {
+        }
+
+        public unsafe void Serialize(ref MessagePackWriter writer, sbyte[] value, MessagePackSerializerOptions options)
+        {
+            if (value == null)
+            {
+                writer.WriteNil();
+                return;
+            }
+
+            var inputLength = value.Length;
+            writer.WriteArrayHeader(inputLength);
+            if (inputLength == 0)
+            {
+                return;
+            }
+
+            fixed (sbyte* pSource = &value[0])
+            {
+                var inputEnd = pSource + inputLength;
+                var inputIterator = pSource;
+
+#if HARDWARE_INTRINSICS_X86
+                const int Stride = 16;
+                if (Popcnt.IsSupported && inputLength >= Stride)
+                {
+                    {
+                        // Make InputIterator Aligned
+                        var offset = UnsafeMemoryAlignmentUtility.CalculateDifferenceAlign16(inputIterator);
+                        inputLength -= offset;
+                        var offsetEnd = inputIterator + offset;
+                        while (inputIterator != offsetEnd)
+                        {
+                            writer.Write(*inputIterator++);
+                        }
+                    }
+
+                    ReadOnlySpan<byte> shuffle = new byte[]
```

Comment:
Is this array mutated or can we allocate it as a `static readonly` field? It's a huge array and we like to be as close to alloc-free as possible during serialization.

Reply:
This is alloc-free code in Realease build.
I will move the `shuffle` table to another file( which is generated by T4).

Reply:
[The sharplab.io sample code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4fsJToFkAFAEo2HLhIBKMAIYATAPIA7XAE8AygAcZSgDzBVGGAD46qmQCsArlboBeOkpgB3OgaMBtALriJndpR+QXSkaL7BnKHhEejRwQCsYYERXFHJKXSx6SmkAJxJGZH5cUEAHGmFdAAMJXQAvgDctbUAZtCyYAAWdIIAbjJQjnTYSmaWNsK1AZWEeYJKwk3ZEnXhq5R1QA).

Since C#7.2(or 7.3 I don't remember...), the Roslyn compiler in Release Build avoids  the `byte[]` allocation when that `byte[]` is static and readonly.

Reply:
Wow. That's very cool. Thanks for sharing.

Reply:
However, I guess we need to know whether it's also alloc-free when Unity compiles it (in each of its many modes). It has an older C# compiler. We wouldn't want to regress that either.

Reply:
Unity doesn't support .NET Core 3.1.
`HARDWARE_INTRINSICS_X86` is never defined.
So you don't have to mind that.

Unity2018.4 and later uses Roslyn compiler and enables this optimization.